gmqtt/message_queue.py

```diff
@@ -1,7 +1,79 @@
+import asyncio
+import enum
+import functools
+import collections
 
-from typing import Litteral
+# from .aioclient import Subscription
+from .client import Message
+
+from typing import List, Tuple, Dict, Callable
+
+
+class TopicFilter:
+    def __init__(self, topic_filter: str):
+        self.levels = topic_filter.split("/")
+        TopicFilter.validate_topic(self.levels)
+
+    @staticmethod
+    def validate_topic(filter_levels: List[str]):
+        if filter_levels[-1][-1] == "#":
+            if len(filter_levels[-1]) > 1:
+                raise ValueError("Multi-level wildcard must be on its own level")
+
+        if "#" in "".join(filter_levels[:-1]):
+            raise ValueError(
+                "Multi-level wildcard must be at the end of the topic filter"
+            )
+
+        for level in filter_levels:
+            if len(level) > 1:
+                if "+" in level:
+                    raise ValueError("Single-level wildcard (+) only allowed by itself")
+
+    def match(self, topic: str) -> bool:
```

Comment:
What do you think about using regexp ?
For example if a user passes a topic ("root/sub1/+/sub3") we can transform it in regexp ("root/sub1/([^/]+)/sub3") and code will be more clear.

Replacements: `+` -> `([^/]+)`, `#` ->`(.+)`

Reply:
Yes, I am considering switching this to a regexp - I wonder how much faster it'll be.

I'm also considering using a tree structure to store the subscriptions, which I think will find all matching subscriptions with less work.

Reply:
Tree structure will give us a speed boost if only you will have hundred of subscriptions. For a less then ten (in my opinion) it will be only overhead for iterating and managing this structure in memory. 

But it's interesting and you may implement it and compare with `for-loop and matching by regexp` implementation.